src/test/consensus/Validations_test.cpp

```diff
@@ -257,238 +291,299 @@ class Validations_test : public beast::unit_test::suite
 
-                // Confirm old ledger on hand, but not new ledger
-                BEAST_EXPECT(
-                    harness.vals().numTrustedForLedger(Ledger::ID{2}) == 1);
-                BEAST_EXPECT(
-                    harness.vals().numTrustedForLedger(Ledger::ID{20}) == 0);
+            // Test the node changing signing key
 
-                // Issue a new signing key and re-issue the validation with a
-                // new ID but the same sequence number
-                a.advanceKey();
+            // Confirm old ledger on hand, but not new ledger
+            BEAST_EXPECT(
+                harness.vals().numTrustedForLedger(ledgerAB.id()) == 1);
+            BEAST_EXPECT(
+                harness.vals().numTrustedForLedger(ledgerABC.id()) == 0);
 
-                // No validations following ID{2}
-                BEAST_EXPECT(harness.vals().getNodesAfter(Ledger::ID{2}) == 0);
+            // Rotate signing keys
+            n.advanceKey();
 
-                BEAST_EXPECT(
-                    AddOutcome::sameSeq == harness.add(a, Ledger::Seq{2}, Ledger::ID{20}));
+            harness.clock().advance(1s);
 
-                // Old ID should be gone ...
-                BEAST_EXPECT(harness.vals().numTrustedForLedger(Ledger::ID{2}) == 0);
-                BEAST_EXPECT(harness.vals().numTrustedForLedger(Ledger::ID{20}) == 1);
-                {
-                    // Should be the only trusted for ID{20}
-                    auto trustedVals =
-                        harness.vals().getTrustedForLedger(Ledger::ID{20});
-                    BEAST_EXPECT(trustedVals.size() == 1);
-                    BEAST_EXPECT(trustedVals[0].key() == a.currKey());
-                    // ... and should be the only node after ID{2}
-                    BEAST_EXPECT(harness.vals().getNodesAfter(Ledger::ID{2}) == 1);
+            // Cannot re-do the same full validation sequence
+            BEAST_EXPECT(
+                ValStatus::badSeq == harness.add(n.validate(ledgerAB)));
+            // Cannot send the same partial validation sequence
+            BEAST_EXPECT(
+                ValStatus::badSeq == harness.add(n.partial(ledgerAB)));
 
-                }
+            // Now trusts the newest ledger too
+            harness.clock().advance(1s);
+            BEAST_EXPECT(
+                ValStatus::current == harness.add(n.validate(ledgerABC)));
+            BEAST_EXPECT(
+                harness.vals().numTrustedForLedger(ledgerAB.id()) == 1);
+            BEAST_EXPECT(
+                harness.vals().numTrustedForLedger(ledgerABC.id()) == 1);
 
-                // A new key, but re-issue a validation with the same ID and
-                // Sequence
-                a.advanceKey();
+            // Processing validations out of order should ignore the older
+            // validation
+            harness.clock().advance(2s);
+            auto const val3 = n.validate(ledgerABCDE);
```

Comment:
nit: different names for `val3` and `val4`

Reply:
I'll go with `valABCDE` and `valABCD` since otherwise its slightly confusing whether `3` or `4` refers to the order in which they were validated, or their relative historical order.